src/counters.ts

```diff
@@ -0,0 +1,694 @@
+
+	// Scheme for moving averages. 
+	@IsNotEmpty()
+	@IsBoolean()
+	public readonly dumpToFile: boolean;
+
+	/**
+	 * whether we should dump the counters to charts identified by the name of the counter. 
+	 *
+	 * @type {boolean}
+	 * @memberof Counters
+	 */
+	@IsNotEmpty()
+	@IsBoolean()
+	public readonly dumpToChart: boolean;
+
+
+	/**
+	 * Name of this counter. This is used to identify all generated files.
+	 *
+	 * @type {string}
+	 * @memberof Counters
+	 */
+	@IsString()
+	@IsNotEmpty()
+	@Matches(/[a-z0-9_]/i, 'i')
+	public readonly name: string;
+
+	/**
+	 * directory where to dump files and charts. default is process's cwd.
+	 *
+	 * @type {string}
+	 * @memberof Counters
+	 */
+	@IsString()
+	@IsNotEmpty()
+	public readonly outputDirectory: string;
+
+	/**
+	 * the collection of counters for this object corresponding to {@link pid}, and {@link includeParent}
+	 *
+	 * @type {Map<number,ProcessStatisticsCollection>}
+	 * @memberof Counters
+	 */
+	public collection: Map<number, ProcessStatisticsCollection> = new Map<number, ProcessStatisticsCollection>();
+
+	/**yarn
+	 * the simple moving average over 4 elements of {@link collection} of counters for this object corresponding to {@link pid}, and {@link includeParent}
+	 *
+	 * @type {Map<number,ProcessStatisticsCollection>}
+	 * @memberof Counters
+	 */
+	public smaOver4Collection: Map<number, ProcessStatisticsCollection> = new Map<number, ProcessStatisticsCollection>();
+
+	/**
+	 * the exponential moving average over 4 elements of {@link collection} of counters for this object corresponding to {@link pid}, and {@link includeParent}
+	 *
+	 * @type {Map<number,ProcessStatisticsCollection>}
+	 * @memberof Counters
+	 */
+	public emaOver4Collection: Map<number, ProcessStatisticsCollection> = new Map<number, ProcessStatisticsCollection>();
+
+	/**
+	 * the computed statistics on the collected results. These are consistent with the number we need to submit to performance frameworks for baselining the results.
+	 *
+	 * @type ComputedStatistics}
+	 * @memberof Counters
+	 */
+	public computedStatistics: ComputedStatistics;
+
+	/**
+	  * Constructor allows for construction with a bunch of optional parameters
+	  *
+	  * 
+	  * @param name - Name of this counter. This is used to identify all generated files.
+	  * @param pid -  the root pid of processes that we are tracking
+	  * @param includeParent - flag to indicate if we should include parent of the pid in the process that we are tracking. Including parent implies we track all of the parent's children and not just the process designated by {@link pid}
+	  * @param object of @type {CountersOptions} with:
+			* @param collectionInterval - see {@link CountersOptions}.
+			* @param includeMovingAverages - see {@link CountersOptions}.
+			* @param dumpToFile - see {@link CountersOptions}.
+			* @param dumpToChart - see {@link CountersOptions}.
+			* @param outputDirectory - see {@link CountersOptions}. 

```

Comment:
align indentation

Reply:
These were purposely subindented as I am describing subfields of CountersOptions.